p2p/discv5/table.go

```diff
@@ -64,42 +65,45 @@ func newTable(ourID NodeID, ourAddr *net.UDPAddr) *Table {
 	return tab
 }
 
-func (tab *Table) chooseBucketFillTarget() common.Hash {
-	bucketCount := nBuckets
-	for bucketCount > 0 && len(tab.buckets[nBuckets-bucketCount].entries) == 0 {
-		bucketCount--
+const printTable = false
+
+func (tab *Table) chooseBucketRefreshTarget() common.Hash {
+	entries := 0
+	if printTable {
+		fmt.Println()
 	}
-	var bucket int
-	for {
-		// select a target hash that could go into a certain randomly selected bucket
-		// buckets are chosen with an even chance out of the existing ones that contain
-		// less that bucketSize entries, plus a potential new one beyond these
-		bucket = nBuckets - 1 - int(randUint(uint32(bucketCount+1)))
-		if bucket == bucketCount || len(tab.buckets[bucket].entries) < bucketSize {
-			break
+	for i, b := range tab.buckets {
+		entries += len(b.entries)
+		if printTable {
+			for _, e := range b.entries {
+				fmt.Println(i, e.state, e.addr().String(), e.ID.String(), e.sha.Hex())
+			}
 		}
 	}
 
-	// calculate target that has the desired log distance from our own address hash
-	target := tab.self.sha.Bytes()
-	prefix := binary.BigEndian.Uint64(target[0:8])
-	shift := uint(nBuckets - 1 - bucket)
-	if bucket != bucketCount {
-		shift++
+	prefix := binary.BigEndian.Uint64(tab.self.sha[0:8])
+
+	dist := ^uint64(0)
+	entry := int(randUint(uint32(entries + 1)))
+	for _, b := range tab.buckets {
+		if entry < len(b.entries) {
+			n := b.entries[entry]
+			dist = binary.BigEndian.Uint64(n.sha[0:8]) ^ prefix
+			break
+		}
+		entry -= len(b.entries)
 	}
-	var b [8]byte
-	rand.Read(b[:])
-	rnd := binary.BigEndian.Uint64(b[:])
-	rndMask := (^uint64(0)) >> shift
-	addrMask := ^rndMask
-	xorMask := uint64(0)
-	if bucket != bucketCount {
-		xorMask = rndMask + 1
+
+	ddist := ^uint64(0)
+	if dist+dist > dist {
+		ddist = dist
 	}
-	prefix = (prefix&addrMask ^ xorMask) | (rnd & rndMask)
-	binary.BigEndian.PutUint64(target[0:8], prefix)
+	targetPrefix := prefix ^ randUint64n(ddist)
+
+	var target common.Hash
+	binary.BigEndian.PutUint64(target[0:8], targetPrefix)
 	rand.Read(target[8:])
-	return common.BytesToHash(target)
+	return target
```

Comment:
I wanted to do a random refresh target selection that selects closer addresses with a higher probability to fill up all buckets and ensure a healthy Kademlia topology. First I wanted to do this by selecting a certain bucket to refresh, but I realized that this approach had a lot of problems. Now I'm selecting a random node from the table and search in a distance radius of a similar order of magnitude. I believe this approach is both more reliable and effective.